### PR TITLE
Add initial VBnet Coding Guidelines

### PR DESCRIPTION
**Pull request by AdamSpeight2008**
@KathleenDollard 
I've implement a set of coding guidelines, as mentioned in #251

### CODE REVIEW

File: proposals/VBnet-Coding-Guidelines.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
+# VB.net Coding Guidelines
+ 
+ Coding Conventions
+ 
+ * 80 character line limit    
+ Is the recommended maximum length of single line of code.    
+*As seen in a text editor, not in the context of the actual language specification*
+
+ * Use 4 space indentation, not tabs.
```
Comment (sharwell) [inline proposals/VBnet-Coding-Guidelines.md]: 💭 This is established by the .editorconfig for the repository. Duplicating items already included there seems unnecessary.

File: proposals/VBnet-Coding-Guidelines.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
+# VB.net Coding Guidelines
+ 
+ Coding Conventions
+ 
+ * 80 character line limit
```
Comment (sharwell) [inline proposals/VBnet-Coding-Guidelines.md]: ❗️ There are two issues with this:

1. There is no real line length limit in place
2. If there was a line length limit, it wouldn't be 80 characters

I wouldn't be against mentioning a soft limit of 120 characters or more.

File: proposals/VBnet-Coding-Guidelines.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+*As seen in a text editor, not in the context of the actual language specification*
+
+ * Use 4 space indentation, not tabs.
+ 
+ * Use plain code to validation parameters at public boundaries.    
+   * Do not use Contracts or magic helpers.
+   * Use `Debug.Assert()` for checks not needed in release builds. Always include a contextually meaningful message string in your assert to identify failure conditions.
+   * Add assertions to documentation assumptions on non-local program state or parameter values, eg.    
+   	*"At this point in parsing the scanner should have been advanced to a '.' token by the caller".*
+   * Avoid allocations in compiler hot paths.
```
Comment (sharwell) [inline proposals/VBnet-Coding-Guidelines.md]: :memo: This is nested under "Use plain code to validat[e] parameters..." but doesn't appear related

**Comment by KathleenDollard** [comment]
Does Roslyn not already have this?

**Comment by AdamSpeight2008** [comment]
@KathleenDollard Only a passing reference, but not a explicit set of guidelines.

**Comment by gafter** [comment]
I don't think a set of coding guidelines belongs as part of the VB language design repository. I think there is little enough code (e.g. examples in the specification) that we can police style guidelines informally.

File: proposals/VBnet-Coding-Guidelines.md (hunk 2 above)
Comment (DualBrain) [reply proposals/VBnet-Coding-Guidelines.md]: Having both an 80 character line limit and 4 character "tab" seems, to me, to conflict a bit... a lot of code could easily have 5 or more "tabs"...

Namespace...
.Class...
..Function...
...Select Case...
....Case...
.....If...

So 1/4th of the line is eaten up with tabs.

Ouch.

File: proposals/VBnet-Coding-Guidelines.md (hunk 1 above)
Comment (AdamSpeight2008) [reply proposals/VBnet-Coding-Guidelines.md]: @sharwell It doesn't harm to state it in the style guidelines though.

File: proposals/VBnet-Coding-Guidelines.md (hunk 2 above)
Comment (AdamSpeight2008) [reply proposals/VBnet-Coding-Guidelines.md]: @DualBrain I know I personally prefer 2 space character tab, but the `/dotnet/...` prefers 4.
@sharwell My rationale behind 80 character was tradition, but I am ok with 120.

The issue / pr was the stimulate discussion, and tempt the team on producing a preferred style for usage in the repos.  *As I was clashing heads, when submit pull-request due the lack of them*

**Comment by AdamSpeight2008** [comment]
@gafter By formal specifying a preferred style guideline for VB code pull requests, you can refer people to the guidelines. Rather than what can seem an arbitrary choice on the whim of the reviewers. By "formally" have a documentation (that you can point to), we can check ourselves prior to submission. Rather than the team initially focusing on style guideline violations.

File: proposals/VBnet-Coding-Guidelines.md (hunk 2 above)
Comment (DualBrain) [reply proposals/VBnet-Coding-Guidelines.md]: @AdamSpeight2008 I would think that the coding guidelines would be based on the language, not to have to adhere to arbitrary limitations decided because of inherit "limitations" of another language.  I think VB should have 2 character "tabs"... having 4 in C# makes sense given the inherit lack of clarity between start/end constructs.  It makes perfect sense to want to have more spacing... it provides a clear differentiation.  VB on the other hand has clear constructs... thus the spacing of 4 is kind of ridiculous.  I also agree that the line limit should be increased... 120 seems more reasonable (especially if 2 spaces were to be utilized).  Again, lines in VB are long... it has greater verbosity.

**Comment by sharwell** [comment]
@AdamSpeight2008 To the degree possible, the guidelines are:

1. Style of existing code should not be altered except by a core contributor or the original author of the code, _and_ should only be done in cases where the new style is obviously correct per one of the following:
    1. The new style is clearly dictated by a style guide for the project, and the old style is not allowed by the style guide
    2. The new style is clearly consistent with the project as a whole, and the old style is clearly not consistent
    3. The new style is applied by automated code cleanup for a code analysis rule enabled for the project (in this case only one type of change should be applied at a time)
2. For cases where the preferred style is not clear, a project-specific code style rule should be proposed (i.e. it is preferable to implement a rule to perform step 1.3 over manual work for 1.1 or 1.2)
3. New code should follow existing patterns, but code reviewers should not be checking manually style rules except for cases where automated enforcement is enabled (i.e. will break the build) or for flagrant deviations from existing style

**Comment by AdamSpeight2008** [comment]
@sharwell My original point is that the coding guideline are not specifically stated, thus as I see can be enforced. Ie *There's no sign or law stating that I can't*

**Comment by sharwell** [comment]
> My original point is that the coding guideline are not specifically stated

This is generally an acceptable state. It means "follow the existing code where possible, but if you are writing new code and don't follow something exactly then that is typically still acceptable".
